host.go

```diff
@@ -0,0 +1,188 @@
+package f3
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/sim"
+	"golang.org/x/xerrors"
+)
+
+type gpbfthost struct {
+	participant *gpbft.Participant
+	manifest    Manifest
+	gpbft.Signer
+	gpbft.Verifier
+	broadcast func(context.Context, []byte) error
+
+	MessageQueue     chan *gpbft.GMessage
+	SelfMessageQueue chan *gpbft.GMessage //for the future when self messages are async
+
+	log Logger
+
+	timer      *time.Timer
+	runningCtx context.Context
+}
+
+func newHost(id gpbft.ActorID, m Manifest, broadcast func(context.Context, []byte) error,
+	s gpbft.Signer, v gpbft.Verifier, log Logger) (*gpbfthost, error) {
+	h := &gpbfthost{
+		manifest:         m,
+		Signer:           s,
+		Verifier:         v,
+		MessageQueue:     make(chan *gpbft.GMessage, 20),
+		SelfMessageQueue: make(chan *gpbft.GMessage, 20),
+		broadcast:        broadcast,
+
+		log: log,
+	}
+
+	// create a stopped timer
+	h.timer = time.NewTimer(100 * time.Hour)
+	if !h.timer.Stop() {
+		<-h.timer.C
+	}
+
+	log.Infof("starting host for P%d", id)
+	p, err := gpbft.NewParticipant(id, h, gpbft.WithTracer(tracer{log}))
+	if err != nil {
+		return nil, xerrors.Errorf("creating participant: %w", err)
+	}
+	h.participant = p
+	return h, nil
+}
+
+func (h *gpbfthost) Run(ctx context.Context) error {
+	h.runningCtx = ctx
+	time.Sleep(2 * time.Second)
+	err := h.participant.Start()
+	if err != nil {
+		return xerrors.Errorf("starting a participant: %w", err)
+	}
+
+loop:
+	for {
+		select {
+		case msg := <-h.SelfMessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, true)
+		default:
+		}
+		if err != nil {
+			break loop
+		}
+
+		select {
+		case <-h.timer.C:
+			h.log.Infof("alarm fired")
+			err = h.participant.ReceiveAlarm()
+		case msg := <-h.SelfMessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, true)
+		case msg := <-h.MessageQueue:
+			_, err = h.participant.ReceiveMessage(msg, false)
+		case <-ctx.Done():
+			err = ctx.Err()
+		}
+		if err != nil {
+			break loop
+		}
+	}
+	h.log.Errorf("gpbfthost exiting: %+v", err)
+	return err
+}
+
+//
+// internal APIs after
+//
+
+// Returns inputs to the next GPBFT instance.
+// These are:
+// - the EC chain to propose,
+// - the power table specifying the participants,
+// - the beacon value for generating tickets.
+// These will be used as input to a subsequent instance of the protocol.
+// The chain should be a suffix of the last chain notified to the host via
+// ReceiveDecision (or known to be final via some other channel).
+func (h *gpbfthost) GetCanonicalChain() (chain gpbft.ECChain, power gpbft.PowerTable, beacon []byte) {
+	// TODO: this is just a complete fake
+	ts := sim.NewTipSetGenerator(1)
```

Comment:
I understand that this is part of a larger body of work. But I am struggling to see why we are using simulation tipset generator here and am unsure where this is going. 

If the purpose of this is to showcase a network over gossipsub then would it make sense to instead: 
* build an API that runs gPBFT over libp2p.
* write a test that uses that library to run a network with gopssipsub transport.

I am struggling to see what the F3 binary is going to be once the larger body of work this is part of is complete.

Reply:
I'm using a simulation tipset generator here because this is the easiest thing to stub this out while we work out the rest of it.

1. We want a standalone binary at some point to facilitate observation, testing, development against a live network, and so on.
2. Library-level tests will be there once things are more complete. Currently, I expect the code to change so much that they would be a distraction.
3.  We need tooling to set up the bootstrap and test networks, which the f3 tool will facilitate.

Reply:
I also don't want to spend time rebasing a multiple thousand lines PR until things are finished. That is why I'm using stubs.